evm/src/cpu/kernel/interpreter.rs

```diff
@@ -0,0 +1,301 @@
+        self.incr(1);
+    }
+
+    fn run_mul(&mut self) {
+        let x = self.pop();
+        let y = self.pop();
+        self.push(x.overflowing_mul(y).0);
+        self.incr(1);
+    }
+
+    fn run_sub(&mut self) {
+        let x = self.pop();
+        let y = self.pop();
+        self.push(x.overflowing_sub(y).0);
+        self.incr(1);
+    }
+
+    fn run_div(&mut self) {
+        let x = self.pop();
+        let y = self.pop();
+        self.push(x / y);
+        self.incr(1);
+    }
+
+    fn run_mod(&mut self) {
+        let x = self.pop();
+        let y = self.pop();
+        self.push(x % y);
+        self.incr(1);
+    }
+
+    fn run_addmod(&mut self) {
+        let x = U512::from(self.pop());
+        let y = U512::from(self.pop());
+        let z = U512::from(self.pop());
+        let res = (x + y) % z;
+        self.push(U256([0, 1, 2, 3].map(|i| res.0[i])));
+        self.incr(1);
+    }
+
+    fn run_mulmod(&mut self) {
+        let x = self.pop();
+        let y = self.pop();
+        let z = U512::from(self.pop());
+        let res = x.full_mul(y) % z;
+        self.push(U256([0, 1, 2, 3].map(|i| res.0[i])));
+        self.incr(1);
+    }
+
+    fn run_exp(&mut self) {
+        let x = self.pop();
+        let y = self.pop();
+        self.push(x.overflowing_pow(y).0);
+        self.incr(1);
+    }
+
+    fn run_lt(&mut self) {
+        let x = self.pop();
+        let y = self.pop();
+        self.push(if x < y { U256::one() } else { U256::zero() });
+        self.incr(1);
+    }
+
+    fn run_gt(&mut self) {
+        let x = self.pop();
+        let y = self.pop();
+        self.push(if x > y { U256::one() } else { U256::zero() });
+        self.incr(1);
+    }
+
+    fn run_eq(&mut self) {
+        let x = self.pop();
+        let y = self.pop();
+        self.push(if x == y { U256::one() } else { U256::zero() });
+        self.incr(1);
+    }
+
+    fn run_iszero(&mut self) {
+        let x = self.pop();
+        self.push(if x.is_zero() {
+            U256::one()
+        } else {
+            U256::zero()
+        });
+        self.incr(1);
+    }
+
+    fn run_and(&mut self) {
+        let x = self.pop();
+        let y = self.pop();
+        self.push(x & y);
+        self.incr(1);
+    }
+
+    fn run_or(&mut self) {
+        let x = self.pop();
+        let y = self.pop();
+        self.push(x | y);
+        self.incr(1);
+    }
+
+    fn run_xor(&mut self) {
+        let x = self.pop();
+        let y = self.pop();
+        self.push(x ^ y);
+        self.incr(1);
+    }
+
+    fn run_not(&mut self) {
+        let x = self.pop();
+        self.push(!x);
+        self.incr(1);
+    }
+
+    fn run_pop(&mut self) {
+        self.pop();
+        self.incr(1);
+    }
+
+    fn run_jump(&mut self) {
+        let x = self.pop().as_usize();
+        self.offset = x;
+        if let Some(&landing_opcode) = self.code.get(self.offset) {
+            assert_eq!(landing_opcode, 0x5b, "Destination is not a JUMPDEST.");
```

Comment:
Hm technically we're supposed to also check that the 0x5b isn't inside some PUSH data, but it's nontrivial and I guess we don't need to strictly conform here

Reply:
Interesting, I didn't know this was also checked. Shouldn't be too hard to add with a preprocessing step. I'll do it in a later PR.